google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/SchemaCompact.java

```diff
@@ -96,4 +118,108 @@ public void check(String tableName, Descriptors.Descriptor userSchema)
               + userSchema.getFields().size());
     }
   }
+
+  /**
+   * @param field
+   * @return True if fieldtype is supported by BQ Schema
+   */
+  public static boolean isSupportedType(Descriptors.FieldDescriptor field) {
+    Descriptors.FieldDescriptor.Type fieldType = field.getType();
+    if (!SupportedTypes.contains(fieldType)) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Method that checks for proper nesting (no recursive protos) and supported types.
+   *
+   * @param message
+   * @param level Keeps track of current level of nesting.
+   * @param allMessageTypes Keeps track of all message types seen, and make sure they don't repeat
+   *     as recursive protos are not supported.
+   * @return True if fieldtype is supported by BQ Schema
+   * @throws IllegalArgumentException if message is invalid
+   */
+  private static boolean isNestedMessageAccepted(
+      Descriptors.Descriptor message, HashSet<Descriptors.Descriptor> allMessageTypes)
+      throws IllegalArgumentException {
+
+    if (allMessageTypes.size() > 15) {
+      return false;
+    }
+
+    if (allMessageTypes.contains(message)) {
+      return false;
+    }
+    allMessageTypes.add(message);
+    boolean result = isSupportedImpl(message, allMessageTypes);
+    allMessageTypes.remove(message);
+    return result;
+  }
+
+  /**
+   * Actual implementation that checks if a user schema is supported.
+   *
+   * @param field
+   * @param level Keeps track of current level of nesting.
+   * @param allMessageTypes Keeps track of all message types seen, and make sure they don't repeat
+   *     as recursive protos are not supported.
+   * @return True if field type and option is supported
+   * @throws IllegalArgumentException if schema is invalid
+   */
+  private static boolean isSupportedImpl(
+      Descriptors.Descriptor userSchema, HashSet<Descriptors.Descriptor> allMessageTypes)
+      throws IllegalArgumentException {
+
+    List<Descriptors.OneofDescriptor> oneofs = userSchema.getOneofs();
```

Comment:
Sorry, the doc is wrong. Oneof can be supported by flattening the fields out. Let's just leave it out of the check.